pkg/urbit/vere/king.c

```diff
@@ -231,54 +236,161 @@ _king_curl_alloc(void* dat_v, size_t uni_t, size_t mem_t, void* buf_v)
   return siz_t;
 }
 
-/* _king_get_atom(): HTTP GET url_c, produce the response body as an atom.
+/* _king_curl_bytes(): HTTP GET url_c, produce response body bytes.
 **  XX deduplicate with dawn.c
 */
-static u3_noun
-_king_get_atom(c3_c* url_c)
+static c3_i
+_king_curl_bytes(c3_c* url_c, c3_w* len_w, c3_y** hun_y)
 {
-  CURL *curl;
-  CURLcode result;
-  long cod_l;
-
+  c3_i     ret_i = 0;
+  CURL    *cul_u;
+  CURLcode res_i;
+  long     cod_i;
   uv_buf_t buf_u = uv_buf_init(c3_malloc(1), 0);
 
-  if ( !(curl = curl_easy_init()) ) {
+  if ( !(cul_u = curl_easy_init()) ) {
     u3l_log("failed to initialize libcurl\n");
     exit(1);
   }
 
-  u3K.ssl_curl_f(curl);
-  curl_easy_setopt(curl, CURLOPT_URL, url_c);
-  curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, _king_curl_alloc);
-  curl_easy_setopt(curl, CURLOPT_WRITEDATA, (void*)&buf_u);
+  u3K.ssl_curl_f(cul_u);
+  curl_easy_setopt(cul_u, CURLOPT_URL, url_c);
+  curl_easy_setopt(cul_u, CURLOPT_WRITEFUNCTION, _king_curl_alloc);
+  curl_easy_setopt(cul_u, CURLOPT_WRITEDATA, (void*)&buf_u);
 
-  result = curl_easy_perform(curl);
-  curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &cod_l);
+  res_i = curl_easy_perform(cul_u);
+  curl_easy_getinfo(cul_u, CURLINFO_RESPONSE_CODE, &cod_i);
 
   //  XX retry?
   //
-  if ( CURLE_OK != result ) {
-    u3l_log("failed to fetch %s: %s\n",
-            url_c, curl_easy_strerror(result));
-    u3_king_bail();
-    exit(1);
+  if ( CURLE_OK != res_i ) {
+    u3l_log("curl: failed %s: %s\n", url_c, curl_easy_strerror(res_i));
+    ret_i = -1;
+  }
+  if ( 300 <= cod_i ) {
+    u3l_log("curl: error %s: HTTP %ld\n", url_c, cod_i);
+    ret_i = -2;
   }
-  if ( 300 <= cod_l ) {
-    u3l_log("error fetching %s: HTTP %ld\n", url_c, cod_l);
+
+  curl_easy_cleanup(cul_u);
+
+  *len_w = buf_u.len;
+  *hun_y = (c3_y*)buf_u.base;
```

Comment:
Can we check that these aren't NULL before writing them? This isn't a hot codepath, and so the extra peace of mind is worth it in my opinion.

Reply:
The purpose of this function is get an HTTP response body as a byte buffer. Why should either the base pointer or length out-params be optional?